# Probability Models


In [1]:
### headers
import numpy as np
from numpy import random

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

```{contents}
:local:
:depth: 2
```


This section introduces probability language and the table structure used to describe data. We start with a **probability model** for a business process: the possible outcomes, their chances, and what we can compute from them. Then we turn to observed data and separate the population or process we care about from the sample we observe, and identify variables, observations, and distributions.


## Probability Models

A **probability model** (or *chance model*) describes a random process by listing every possible **outcome** and the **probability** of each. The list of all possible outcomes is the **sample space**.

Consider an online store that ships orders. When a shipment goes out, the store does not know whether it will arrive on time. From past experience, the operations team uses this model for one shipment:


In [2]:
shipping = pd.DataFrame({
    'outcome': ['on time', '1 day late', '3+ days late'],
    'probability': [0.80, 0.15, 0.05],
})
shipping


,outcome,probability
0,on time,0.80
1,1 day late,0.15
2,3+ days late,0.05


Every probability model follows two rules:

1. Each probability is between 0 (impossible) and 1 (certain).
2. The probabilities of all the outcomes in the sample space add up to 1, because exactly one of them must happen.


In [3]:
shipping['probability'].sum()


np.float64(1.0)

### Events and Complements

An **event** is a set of outcomes we care about. The event "the shipment is late" contains two outcomes, `1 day late` and `3+ days late`, so its probability is the sum of their probabilities.


In [4]:
late = shipping['outcome'] != 'on time'
p_late = shipping.loc[late, 'probability'].sum()
p_late


np.float64(0.2)

The **complement** of an event is "the event does not happen." Because the probabilities add up to 1, the probability of the complement is 1 minus the probability of the event:

$$P(\text{not late}) = 1 - P(\text{late})$$

The complement is often the easier way to compute a probability: "at least one day late" is everything except "on time."


In [5]:
p_not_late = 1 - p_late
p_not_late


np.float64(0.8)

### Expected Value

Managers usually care about a number attached to each outcome, such as a cost. Suppose the store gives a \$5 credit when a shipment is 1 day late and a \$20 credit when it is 3 or more days late. The credit is a **random variable**: a numeric quantity whose value depends on chance.

The **expected value** of a random variable is the sum of each value times its probability:

$$E(\text{credit}) = 0 \times 0.80 + 5 \times 0.15 + 20 \times 0.05$$


In [6]:
shipping['credit'] = [0, 5, 20]
expected_credit = (shipping['credit'] * shipping['probability']).sum()
expected_credit


np.float64(1.75)

The expected credit is \$1.75 per shipment. No single shipment costs \$1.75: each costs \$0, \$5, or \$20. The expected value is the **long-run average** over many shipments. If the store ships 10,000 orders a month, it should budget about \$17,500 a month for late-delivery credits.

### From Model to Data

A probability model describes the process; data are what we actually observe. We can connect the two with a **simulation**: use the model to generate many shipments and look at what happens. `rng.choice` draws outcomes at random with the model's probabilities.


In [7]:
rng = np.random.default_rng(8)
simulated = pd.Series(rng.choice(shipping['outcome'], size=10000, p=shipping['probability']))

simulated.value_counts(normalize=True)


on time         0.8016
1 day late      0.1525
3+ days late    0.0459
Name: proportion, dtype: float64

The proportions in the simulated data, the **empirical distribution**, are close to the model's probabilities, but not exactly equal: each simulation gives slightly different results. The same is true of the average credit:


In [8]:
credit_by_outcome = shipping.set_index('outcome')['credit']
simulated.map(credit_by_outcome).mean()


np.float64(1.6805)

The simulated average credit is close to the expected value of \$1.75. With more shipments, it tends to get closer. The next section returns to this idea as the law of large numbers.

In practice, we usually work the other way around: we observe data and use them to learn about the process that produced them. The rest of this section introduces the language for describing observed data.


## Probability, Data, and Variables

When describing data, make sure you cover center, spread, and shape.


In [9]:
tips = sns.load_dataset('tips')
tips                      ### size means number of ppl in the party

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4
...,...,...,...,...,...,...,...
239,29.03,5.92,Male,No,Sat,Dinner,3
240,27.18,2.00,Female,Yes,Sat,Dinner,2
241,22.67,2.00,Male,Yes,Sat,Dinner,2
242,17.82,1.75,Male,No,Sat,Dinner,2


In [10]:
tips.shape       ### a dataframe has shape, which is a tuple

(244, 7)

### Population vs Sample
Population is the full group we care about (e.g., “all lunch bills at the restaurant this year”), whereas sample is the subset we actually measured (e.g., the 244 rows in the tips dataset). We only ever see the sample, but we talk like we know the population.
  
### Variable vs Observation
Each **row** in the dataframe/dataset is an observation (e.g., one restaurant bill), whereas **columns** are variables (e.g., total_bill, tip, smoker, time (Dinner/Lunch), etc.).

- **Quantitative** variable: e.g., total_bill (numeric, measured)
- **Categorical** variable: time (Lunch vs Dinner), smoker (Yes/No)

### Distribution

A **distribution** shows how often different values occur. For numeric data:
- Are most bills around \$10? \$20? \$50?
- Are there a few very large bills? (outliers)

For categorical data:
- How common is “Dinner” vs “Lunch”?



## Section Practice

Use this short exercise to check the main workflow from this section.


In [11]:
### Exercise: Practice Probability Models
# A bakery models how many custom cakes are ordered in a day.
#   1. Check that the probabilities add up to 1.
#   2. Compute the probability of at least one order, using the complement of 0 orders.
#   3. Compute the expected number of orders per day.
# Print exactly:
# Total probability: 1.0
# P(at least one order): 0.9
# Expected orders: 1.7
demand = [0, 1, 2, 3]                       # cakes ordered in a day
probabilities = [0.10, 0.30, 0.40, 0.20]
### Your code starts here




### Your code ends here


In [12]:
### Solution
demand = [0, 1, 2, 3]                       # cakes ordered in a day
probabilities = [0.10, 0.30, 0.40, 0.20]

total = round(sum(probabilities), 2)
p_at_least_one = round(1 - probabilities[0], 2)
expected = round(sum(d * p for d, p in zip(demand, probabilities)), 2)

print('Total probability:', total)
print('P(at least one order):', p_at_least_one)
print('Expected orders:', expected)


Total probability: 1.0
P(at least one order): 0.9
Expected orders: 1.7
